File: adventofcode/year2015/day05.py
from adventofcode import Solution

import re


class Day05(Solution):
    def __init__(self, year: str, day: str):
        super().__init__(year, day)

        self._strings = list(map(lambda l: l.strip(), self._load_input_as_lines()))

    def part_one(self):
        return len([i for i in self._strings if len(re.findall(r"(ab|cd|pq|xy)", i)) == 0 and len(re.findall(r"[aeiou]", i)) >= 3 and len(re.findall(r"(?P<letter>[a-z])(?P=letter)", i)) >= 1])

    def part_two(self):
        return len([i for i in self._strings if len(re.findall(r"(?P<letter>[a-z][a-z]).*(?P=letter)", i)) >= 1 and len(re.findall(r"(?P<letter>[a-z])[a-z](?P=letter)", i)) >= 1])

File: adventofcode/year2017/Day05.cs
using adventofcode.common;

namespace adventofcode.year2017;

public class Day05 : Solution
{
    private string[] _input;
    
    public Day05(string year, string day) : base(year, day)
    {
        _input = LoadInputAsLines();
    }

    public override string PartOne()
    {
        var instructions = _input.Select(i => Convert.ToInt32(i)).ToArray();
        var steps = 0;
        var index = 0;

        while (index < instructions.Length)
        {
            index += instructions[index]++;
            steps++;
        }

        return Convert.ToString(steps);
    }

    public override string PartTwo()
    {
        var instructions = _input.Select(i => Convert.ToInt32(i)).ToArray();
        var steps = 0;
        var index = 0;
        var nextIndex = index;

        while (index < instructions.Length)
        {
            var offset = instructions[index];
            nextIndex += offset;
            instructions[index] += offset >= 3 ? -1 : 1;
            
            steps++;
            index = nextIndex;
        }

        return Convert.ToString(steps);
    }
}